boutdata/restart.py

```diff
@@ -184,19 +184,26 @@ def is_pow2(x):
                     break
 
             nx, ny, nz = data.shape
+            dx, dy, dz = old['dx'].flat[0], old['dy'].flat[0], old['dz'].flat[0]
+            # shift grid if CELL-CENTRED   
+            shift = 0.5 if old.attributes(var)['cell_location'] == 'CELL-CENTRE' else 0
+            
             # Make coordinates
             # NOTE: The max min of the coordinates are irrelevant when
             #       interpolating (as long as old and new coordinates
             #       are consistent), so we just choose all variable to
-            #       be between 0 and 1 Calculate the old coordinates
-            xCoordOld = np.linspace(0, 1, nx)
-            yCoordOld = np.linspace(0, 1, ny)
-            zCoordOld = np.linspace(0, 1, nz)
-
+            #       be between 0 and 1 Calculate the old coordinates 
+            xCoordOld = np.arange(nx - mxg + shift)*dx
+            yCoordOld = np.arange(ny - myg + shift)*dy
+            zCoordOld = np.arange(nz + shift)*dz
```

Comment:
@mredenti have you tested this? It doesn't look to me like it would work:
* `np.arange()` as used here will always return integer values
* both the 'old' and 'new' coordinates will have the wrong size - i.e. different number of points from the corresponding dimension in the old/new data array.

I think you need to specify both the `start` and `stop` arguments to `np.arange()`...

Reply:
> @mredenti have you tested this? It doesn't look to me like it would work:
> 
> * `np.arange()` as used here will always return integer values
> * both the 'old' and 'new' coordinates will have the wrong size - i.e. different number of points from the corresponding dimension in the old/new data array.
> 
> I think you need to specify both the `start` and `stop` arguments to `np.arange()`...

@johnomotani You are absolutely right. I made a typo when copying from my tests and the parenthesis should be outside. Apologies for this, I will fix it right away

Reply:
That is a good point. In boundary or guard cells we could just not worry about the NaNs as they should be replaced by communication or boundary conditions before being used. For the z-direction though there are no boundary or guard cells, so we would end up with NaNs on grid points. As far as I remember there is no way to tell `scipy`'s interpolation routines that some dimension is periodic. It would be possible to fix the affected points by doing the interpolation 'by hand' using periodicity, but I think just extrapolating is good enough. TL;DR I think we should just pass `kwargs : {'fill_value' : 'extrapolate'}` all the time.

For multiple restart files, I the function works as documented for multiple files. The only issue is that `newNx`, `newNy` and `newNz` are a bit inconvenient - at the moment you have to pass values that depend on how many restart files you have. I'd prefer for `newNx`, `newNy` and `newNz` to accept the same values that would be in the `BOUT.inp` for the new simulation, which would require reading the original global grid sizes and using those to calculate the per-file new grid sizes from the global inputs; would also require some tweaks to the logic for when boundary points should or should not be included (see https://bout-dev.readthedocs.io/en/latest/user_docs/bout_options.html#grids).

Reply:
> That is a good point. In boundary or guard cells we could just not worry about the NaNs as they should be replaced by communication or boundary conditions before being used. 

In other words, it does not really matter what values end up on the boundary guard cells and ghost guard cells since BOUT++ will enforce the correct values before any derivative evaluation?

>For the z-direction though there are no boundary or guard cells, so we would end up with NaNs on grid points. As far as I remember there is no way to tell `scipy`'s interpolation routines that some dimension is periodic. It would be possible to fix the affected points by doing the interpolation 'by hand' using periodicity, but I think just extrapolating is good enough. TL;DR I think we should just pass `kwargs : {'fill_value' : 'extrapolate'}` all the time.

Yes, `scipy`'s interpolation routine does not support interpolation with periodic boundaries which is a shame really. I guess we could pad the sides with a copy of the data on the opposite side/row but it becomes quite fiddly. 

> For multiple restart files, I the function works as documented for multiple files. The only issue is that `newNx`, `newNy` and `newNz` are a bit inconvenient - at the moment you have to pass values that depend on how many restart files you have. I'd prefer for `newNx`, `newNy` and `newNz` to accept the same values that would be in the `BOUT.inp` for the new simulation, which would require reading the original global grid sizes and using those to calculate the per-file new grid sizes from the global inputs; would also require some tweaks to the logic for when boundary points should or should not be included (see https://bout-dev.readthedocs.io/en/latest/user_docs/bout_options.html#grids).

This would be a nice feature to add. So I need to be careful with the following 

`nx is the number of points in X including the boundaries`, 
`ny and nz are the number of points in Y and Z not including the boundaries`
and the fact that `boutdata.restart.resize` accepts a `MXG` and `MYG` parameter

Reply:
@mredenti yes to all of that :+1:

Reply:
The commit b98f7357e0ebca1833561a5db0cd2615ef086467 takes care of extrapolation if the grid points to be interpolated are outside the range. It turns out one had to pass the argument `fill_value = None`  to the RegularGridInterpolator. The default fill_value is np.nan.